src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlBuffer.cs

```diff
@@ -1309,5 +1342,59 @@ private void ThrowIfNull()
                 throw new SqlNullValueException();
             }
         }
+        // [Field]As<T> method explanation:
+        // these methods are used to bridge generic to non-generic access to value type fields on the storage struct
+        // where typeof(T) == typeof(field) 
+        //   1) RyuJIT will recognize the pattern of (T)(object)T as being redundant and eliminate 
+        //   the T and object casts leaving T, so while this looks like it will put every value type instance in a box the 
+        //   generated assembly will be short and direct
+        //   2) another jit may not recognize the pattern and should emit the code as seen. this will box and then unbox the
+        //   value type which is no worse than the mechanism that this code replaces
+        // where typeof(T) != typeof(field)
+        //   the jit will emit all the cast operations as written. this will put the value into a box and then attempt to
+        //   cast it, because it is an object no conversions are used and this will generate the desired InvalidCastException       
+        //   for example users cannot widen a short to an int preserving external expectations 
+
+        internal T ByteAs<T>()
+        {
+            ThrowIfNull();
+            return (T)(object)_value._byte;
+        }
+
+        internal T BooleanAs<T>()
+        {
+            ThrowIfNull();
+            return (T)(object)_value._boolean;
+        }
+
+        internal T Int32As<T>()
+        {
+            ThrowIfNull();
+            return (T)(object)_value._int32;
+        }
+
+        internal T Int16As<T>()
+        {
+            ThrowIfNull();
+            return (T)(object)_value._int16;
+        }
+
+        internal T Int64As<T>()
+        {
+            ThrowIfNull();
+            return (T)(object)_value._int64;
+        }
+
+        internal T DoubleAs<T>()
+        {
+            ThrowIfNull();
+            return (T)(object)_value._double;
+        }
+
+        internal T SingleAs<T>()
+        {
+            ThrowIfNull();
+            return (T)(object)_value._single;
+        }
```

Comment:
@Wraith2 I think you have these changes in two places 
https://github.com/dotnet/SqlClient/pull/1348/files

Reply:
@JRahnama Other PR is approved, if we can merge that it will be easier to resolve conflicts of that PR in here.

Reply:
rebased to current master , that should resolve this.